File: src/libs/m3/Syscalls.cc
/*
 * Copyright (C) 2015, Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#include <base/log/Lib.h>
#include <base/tracing/Tracing.h>
#include <base/Errors.h>
#include <base/Init.h>

#include <m3/com/GateStream.h>
#include <m3/Syscalls.h>

namespace m3 {

INIT_PRIO_SYSC Syscalls Syscalls::_inst;

DTU::Message *Syscalls::send_receive(const void *msg, size_t size) {
    DTU::get().send(_gate.ep(), msg, size, 0, m3::DTU::SYSC_REP);

    DTU::Message *reply;
    do {
        DTU::get().try_sleep(false);

        reply = DTU::get().fetch_msg(m3::DTU::SYSC_REP);
    }
    while(reply == nullptr);

    return reply;
}

Errors::Code Syscalls::send_receive_result(const void *msg, size_t size) {
    DTU::Message *reply = send_receive(msg, size);

    KIF::DefaultReply *rdata = reinterpret_cast<KIF::DefaultReply*>(reply->data);
    Errors::last = static_cast<Errors::Code>(rdata->error);

    DTU::get().mark_read(m3::DTU::SYSC_REP, reinterpret_cast<size_t>(reply));
    return Errors::last;
}

Errors::Code Syscalls::createsrv(capsel_t dst, capsel_t rgate, const String &name) {
    LLOG(SYSC, "createsrv(dst=" << dst << ", rgate=" << rgate << ", name=" << name << ")");

    KIF::Syscall::CreateSrv req;
    req.opcode = KIF::Syscall::CREATE_SRV;
    req.dst_sel = dst;
    req.rgate_sel = rgate;
    req.namelen = Math::min(name.length(), sizeof(req.name));
    memcpy(req.name, name.c_str(), req.namelen);
    size_t msgsize = sizeof(req) - sizeof(req.name) + req.namelen;
    return send_receive_result(&req, Math::round_up(msgsize, DTU_PKG_SIZE));
}

Errors::Code Syscalls::createsess(capsel_t dst, capsel_t srv, word_t ident) {
    LLOG(SYSC, "createsessat(dst=" << dst << ", srv=" << srv << ", ident=" << fmt(ident, "0x") << ")");

    KIF::Syscall::CreateSess req;
    req.opcode = KIF::Syscall::CREATE_SESS;
    req.dst_sel = dst;
    req.srv_sel = srv;
    req.ident = ident;
    return send_receive_result(&req, sizeof(req));
}

Errors::Code Syscalls::creatergate(capsel_t dst, int order, int msgorder) {
    LLOG(SYSC, "creatergate(dst=" << dst << ", size=" << fmt(1UL << order, "x")
        << ", msgsize=" << fmt(1UL << msgorder, "x") << ")");

    KIF::Syscall::CreateRGate req;
    req.opcode = KIF::Syscall::CREATE_RGATE;
    req.dst_sel = dst;
    req.order = static_cast<xfer_t>(order);
    req.msgorder = static_cast<xfer_t>(msgorder);
    return send_receive_result(&req, sizeof(req));
}

Errors::Code Syscalls::createsgate(capsel_t dst, capsel_t rgate, label_t label, word_t credits) {
    LLOG(SYSC, "createsgate(dst=" << dst << ", rgate=" << rgate << ", label=" << fmt(label, "#x")
        << ", credits=" << credits << ")");

    KIF::Syscall::CreateSGate req;
    req.opcode = KIF::Syscall::CREATE_SGATE;
    req.dst_sel = dst;
    req.rgate_sel = rgate;
    req.label = label;
    req.credits = credits;
    return send_receive_result(&req, sizeof(req));
}

Errors::Code Syscalls::createmgate(capsel_t dst, goff_t addr, size_t size, int perms) {
    LLOG(SYSC, "createmgate(dst=" << dst << ", addr=" << addr << ", size=" << size
        << ", perms=" << perms << ")");

    KIF::Syscall::CreateMGate req;
    req.opcode = KIF::Syscall::CREATE_MGATE;
    req.dst_sel = dst;
    req.addr = addr;
    req.size = size;
    req.perms = static_cast<xfer_t>(perms);
    return send_receive_result(&req, sizeof(req));
}

Errors::Code Syscalls::createmap(capsel_t dst, capsel_t vpe, capsel_t mgate, capsel_t first,
                                 capsel_t pages, int perms) {
    LLOG(SYSC, "createmap(dst=" << dst << ", vpe=" << vpe << ", mgate=" << mgate
        << ", first=" << first << ", pages=" << pages << ", perms=" << perms << ")");

    KIF::Syscall::CreateMap req;
    req.opcode = KIF::Syscall::CREATE_MAP;
    req.dst_sel = dst;
    req.vpe_sel = vpe;
    req.mgate_sel = mgate;
    req.first = first;
    req.pages = pages;
    req.perms = static_cast<xfer_t>(perms);
    return send_receive_result(&req, sizeof(req));
}

Errors::Code Syscalls::createvpegrp(capsel_t dst) {
    LLOG(SYSC, "createvpegrp(dst=" << dst << ")");

    KIF::Syscall::CreateVPEGrp req;
    req.opcode = KIF::Syscall::CREATE_VPEGRP;
    req.dst_sel = dst;
    return send_receive_result(&req, sizeof(req));
}

Errors::Code Syscalls::createvpe(const KIF::CapRngDesc &dst, capsel_t sgate, const String &name,
                                 PEDesc &pe, epid_t sep, epid_t rep, bool tmuxable, capsel_t group) {
    LLOG(SYSC, "createvpe(dst=" << dst << ", sgate=" << sgate
        << ", name=" << name << ", type=" << static_cast<int>(pe.type())
        << ", sep=" << sep << ", rep=" << rep << ", tmuxable=" << tmuxable
        << ", group=" << group << ")");

    KIF::Syscall::CreateVPE req;
    req.opcode = KIF::Syscall::CREATE_VPE;
    req.dst_crd = dst.value();
    req.sgate_sel = sgate;
    req.pe = pe.value();
    req.sep = sep;
    req.rep = rep;
    req.muxable = tmuxable;
    req.group_sel = group;
    req.namelen = Math::min(name.length(), sizeof(req.name));
    memcpy(req.name, name.c_str(), req.namelen);

    size_t msgsize = sizeof(req) - sizeof(req.name) + req.namelen;
    DTU::Message *msg = send_receive(&req, Math::round_up(msgsize, DTU_PKG_SIZE));
    auto *reply = reinterpret_cast<KIF::Syscall::CreateVPEReply*>(msg->data);

    Errors::last = static_cast<Errors::Code>(reply->error);
    if(Errors::last == Errors::NONE)
        pe = PEDesc(reply->pe);

    DTU::get().mark_read(m3::DTU::SYSC_REP, reinterpret_cast<size_t>(reply));
    return Errors::last;
}

Errors::Code Syscalls::activate(capsel_t ep, capsel_t gate, goff_t addr) {
    LLOG(SYSC, "activate(ep=" << ep << ", gate=" << gate << ")");

    KIF::Syscall::Activate req;
    req.opcode = KIF::Syscall::ACTIVATE;
    req.ep_sel = ep;
    req.gate_sel = gate;
    req.addr = addr;
    return send_receive_result(&req, sizeof(req));
}

Errors::Code Syscalls::vpectrl(capsel_t vpe, KIF::Syscall::VPEOp op, xfer_t arg) {
    LLOG(SYSC, "vpectrl(vpe=" << vpe << ", op=" << op << ", arg=" << arg << ")");

    KIF::Syscall::VPECtrl req;
    req.opcode = KIF::Syscall::VPE_CTRL;
    req.vpe_sel = vpe;
    req.op = static_cast<xfer_t>(op);
    req.arg = arg;
    return send_receive_result(&req, sizeof(req));
}

Errors::Code Syscalls::vpewait(const capsel_t *vpes, size_t count, capsel_t *vpe, int *exitcode) {
    LLOG(SYSC, "vpewait(vpes=" << vpes << ")");

    KIF::Syscall::VPEWait req;
    req.opcode = KIF::Syscall::VPE_WAIT;
    req.vpe_count = count;
    for(size_t i = 0; i < count; ++i)
        req.sels[i] = vpes[i];

    DTU::Message *msg = send_receive(&req, sizeof(req));
    auto *reply = reinterpret_cast<KIF::Syscall::VPEWaitReply*>(msg->data);

    Errors::last = static_cast<Errors::Code>(reply->error);
    if(Errors::last == Errors::NONE) {
        *vpe = reply->vpe_sel;
        *exitcode = reply->exitcode;
    }

    DTU::get().mark_read(m3::DTU::SYSC_REP, reinterpret_cast<size_t>(reply));
    return Errors::last;
}

Errors::Code Syscalls::derivemem(capsel_t dst, capsel_t src, goff_t offset, size_t size, int perms) {
    LLOG(SYSC, "derivemem(dst=" << dst << ", src=" << src << ", off=" << offset
            << ", size=" << size << ", perms=" << perms << ")");

    KIF::Syscall::DeriveMem req;
    req.opcode = KIF::Syscall::DERIVE_MEM;
    req.dst_sel = dst;
    req.src_sel = src;
    req.offset = offset;
    req.size = size;
    req.perms = static_cast<xfer_t>(perms);
    return send_receive_result(&req, sizeof(req));
}

Errors::Code Syscalls::opensess(capsel_t dst, const String &name, xfer_t arg) {
    LLOG(SYSC, "opensess(dst=" << dst << ", name=" << name << ", arg=" << arg << ")");

    KIF::Syscall::OpenSess req;
    req.opcode = KIF::Syscall::OPEN_SESS;
    req.dst_sel = dst;
    req.arg = arg;
    req.namelen = Math::min(name.length(), sizeof(req.name));
    memcpy(req.name, name.c_str(), req.namelen);
    size_t msgsize = sizeof(req) - sizeof(req.name) + req.namelen;
    return send_receive_result(&req, Math::round_up(msgsize, DTU_PKG_SIZE));
}

Errors::Code Syscalls::exchange(capsel_t vpe, const KIF::CapRngDesc &own, capsel_t other, bool obtain) {
    LLOG(SYSC, "exchange(vpe=" << vpe << ", own=" << own << ", other=" << other
        << ", obtain=" << obtain << ")");

    KIF::Syscall::Exchange req;
    req.opcode = KIF::Syscall::EXCHANGE;
    req.vpe_sel = vpe;
    req.own_crd = own.value();
    req.other_sel = other;
    req.obtain = obtain;
    return send_receive_result(&req, sizeof(req));
}

Errors::Code Syscalls::exchangesess(capsel_t vpe, capsel_t sess, const KIF::CapRngDesc &crd,
                                    KIF::ExchangeArgs *args, bool obtain) {
    KIF::Syscall::ExchangeSess req;
    req.opcode = obtain ? KIF::Syscall::OBTAIN : KIF::Syscall::DELEGATE;
    req.vpe_sel = vpe;
    req.sess_sel = sess;
    req.crd = crd.value();
    if(args)
        memcpy(&req.args, args, sizeof(*args));
    else
        req.args.count = 0;

    DTU::Message *msg = send_receive(&req, sizeof(req));
    auto *reply = reinterpret_cast<KIF::Syscall::ExchangeSessReply*>(msg->data);

    Errors::last = static_cast<Errors::Code>(reply->error);
    if(Errors::last == Errors::NONE && args)
        memcpy(args, &reply->args, sizeof(*args));

    DTU::get().mark_read(m3::DTU::SYSC_REP, reinterpret_cast<size_t>(reply));
    return Errors::last;
}

Errors::Code Syscalls::delegate(capsel_t vpe, capsel_t sess, const KIF::CapRngDesc &crd,
                                KIF::ExchangeArgs *args) {
    LLOG(SYSC, "delegate(vpe=" << vpe << ", sess=" << sess << ", crd=" << crd << ")");
    return exchangesess(vpe, sess, crd, args, false);
}

Errors::Code Syscalls::obtain(capsel_t vpe, capsel_t sess, const KIF::CapRngDesc &crd,
                              KIF::ExchangeArgs *args) {
    LLOG(SYSC, "obtain(vpe=" << vpe << ", sess=" << sess << ", crd=" << crd << ")");
    return exchangesess(vpe, sess, crd, args, true);
}

Errors::Code Syscalls::revoke(capsel_t vpe, const KIF::CapRngDesc &crd, bool own) {
    LLOG(SYSC, "revoke(vpe=" << vpe << ", crd=" << crd << ", own=" << own << ")");

    KIF::Syscall::Revoke req;
    req.opcode = KIF::Syscall::REVOKE;
    req.vpe_sel = vpe;
    req.crd = crd.value();
    req.own = own;
    return send_receive_result(&req, sizeof(req));
}

Errors::Code Syscalls::forwardmsg(capsel_t sgate, capsel_t rgate, const void *msg, size_t len,
                                  label_t rlabel, event_t event) {
    LLOG(SYSC, "forwardmsg(sgate=" << sgate << ", rgate=" << rgate << ", msg=" << msg
        << ", len=" << len << ", rlabel=" << fmt(rlabel, "0x") << ", event=" << fmt(event, "0x") << ")");

    KIF::Syscall::ForwardMsg req;
    req.opcode = KIF::Syscall::FORWARD_MSG;
    req.sgate_sel = sgate;
    req.rgate_sel = rgate;
    req.rlabel = rlabel;
    req.event = event;
    req.len = len;
    memcpy(req.msg, msg, Math::min(sizeof(req.msg), len));
    size_t msgsize = sizeof(req) - sizeof(req.msg) + req.len;
    return send_receive_result(&req, Math::round_up(msgsize, DTU_PKG_SIZE));
}

Errors::Code Syscalls::forwardmem(capsel_t mgate, void *data, size_t len, goff_t offset,
                                  uint flags, event_t event) {
    LLOG(SYSC, "forwardmem(mgate=" << mgate << ", data=" << data << ", len=" << len
        << ", offset=" << offset << ", flags=" << fmt(flags, "0x") << ", event=" << fmt(event, "0x") << ")");

    KIF::Syscall::ForwardMem req;
    req.opcode = KIF::Syscall::FORWARD_MEM;
    req.mgate_sel = mgate;
    req.offset = offset;
    req.flags = flags;
    req.event = event;
    req.len = len;
    if(flags & KIF::Syscall::ForwardMem::WRITE)
        memcpy(req.data, data, Math::min(sizeof(req.data), len));

    DTU::Message *msg = send_receive(&req, sizeof(req) - sizeof(req.data) + req.len);
    auto *reply = reinterpret_cast<KIF::Syscall::ForwardMemReply*>(msg->data);

    Errors::last = static_cast<Errors::Code>(reply->error);
    if(Errors::last == Errors::NONE && (~flags & KIF::Syscall::ForwardMem::WRITE))
        memcpy(data, reply->data, len);

    DTU::get().mark_read(m3::DTU::SYSC_REP, reinterpret_cast<size_t>(reply));
    return Errors::last;
}

Errors::Code Syscalls::forwardreply(capsel_t rgate, const void *msg, size_t len,
                                    goff_t msgaddr, event_t event) {
    LLOG(SYSC, "forwardreply(rgate=" << rgate << ", msg=" << msg << ", len=" << len
        << ", msgaddr=" << (void*)msgaddr << ", event=" << fmt(event, "0x") << ")");

    KIF::Syscall::ForwardReply req;
    req.opcode = KIF::Syscall::FORWARD_REPLY;
    req.rgate_sel = rgate;
    req.msgaddr = msgaddr;
    req.event = event;
    req.len = len;
    memcpy(req.msg, msg, Math::min(sizeof(req.msg), len));
    size_t msgsize = sizeof(req) - sizeof(req.msg) + req.len;
    return send_receive_result(&req, Math::round_up(msgsize, DTU_PKG_SIZE));
}

Errors::Code Syscalls::noop() {
    KIF::Syscall::Noop req;
    req.opcode = KIF::Syscall::NOOP;
    return send_receive_result(&req, sizeof(req));
}

// the USED seems to be necessary, because the libc calls it and LTO removes it otherwise
USED void Syscalls::exit(int exitcode) {
    LLOG(SYSC, "exit(code=" << exitcode << ")");

    EVENT_TRACE_FLUSH();

    KIF::Syscall::VPECtrl req;
    req.opcode = KIF::Syscall::VPE_CTRL;
    req.vpe_sel = 0;
    req.op = KIF::Syscall::VCTRL_STOP;
    req.arg = static_cast<xfer_t>(exitcode);
    DTU::get().send(_gate.ep(), &req, sizeof(req), 0, m3::DTU::SYSC_REP);
}

}

File: src/libs/rustm3/src/syscalls.rs
use cap::Selector;
use core::intrinsics;
use dtu;
use errors::Error;
use kif::{CapRngDesc, syscalls, Perm, PEDesc};
use util;

struct Reply<R: 'static> {
    msg: &'static dtu::Message,
    data: &'static R,
}

impl<R: 'static> Drop for Reply<R> {
    fn drop(&mut self) {
        dtu::DTU::mark_read(dtu::SYSC_REP, self.msg);
    }
}

fn send<T>(msg: *const T) -> Result<(), Error> {
    dtu::DTU::send(dtu::SYSC_SEP, msg as *const u8, util::size_of::<T>(), 0, dtu::SYSC_REP)
}

fn send_receive<T, R>(msg: *const T) -> Result<Reply<R>, Error> {
    send(msg)?;

    loop {
        dtu::DTU::try_sleep(false, 0)?;

        let msg = dtu::DTU::fetch_msg(dtu::SYSC_REP);
        if let Some(m) = msg {
            let data: &[R] = unsafe { intrinsics::transmute(&m.data) };
            return Ok(Reply {
                msg: m,
                data: &data[0],
            })
        }
    }
}

fn send_receive_result<T>(msg: *const T) -> Result<(), Error> {
    let reply: Reply<syscalls::DefaultReply> = send_receive(msg)?;

    match reply.data.error {
        0 => Ok(()),
        e => Err(Error::from(e as u32)),
    }
}

pub fn create_srv(dst: Selector, rgate: Selector, name: &str) -> Result<(), Error> {
    log!(
        SYSC,
        "syscalls::create_srv(dst={}, rgate={}, name={})",
        dst, rgate, name
    );

    let mut req = syscalls::CreateSrv {
        opcode: syscalls::Operation::CREATE_SRV.val,
        dst_sel: dst as u64,
        rgate_sel: rgate as u64,
        namelen: name.len() as u64,
        name: unsafe { intrinsics::uninit() },
    };

    // copy name
    for (a, c) in req.name.iter_mut().zip(name.bytes()) {
        *a = c as u8;
    }

    send_receive_result(&req)
}

pub fn activate(ep: Selector, gate: Selector, addr: usize) -> Result<(), Error> {
    log!(
        SYSC,
        "syscalls::activate(ep={}, gate={}, addr={})",
        ep, gate, addr
    );

    let req = syscalls::Activate {
        opcode: syscalls::Operation::ACTIVATE.val,
        ep_sel: ep as u64,
        gate_sel: gate as u64,
        addr: addr as u64,
    };
    send_receive_result(&req)
}

pub fn create_sess(dst: Selector, name: &str, arg: u64) -> Result<(), Error> {
    log!(
        SYSC,
        "syscalls::create_sess(dst={}, name={}, arg={:#x})",
        dst, name, arg
    );

    let mut req = syscalls::CreateSess {
        opcode: syscalls::Operation::CREATE_SESS.val,
        dst_sel: dst as u64,
        namelen: name.len() as u64,
        name: unsafe { intrinsics::uninit() },
        arg: arg,
    };

    // copy name
    for (a, c) in req.name.iter_mut().zip(name.bytes()) {
        *a = c as u8;
    }

    send_receive_result(&req)
}

pub fn create_sgate(dst: Selector, rgate: Selector, label: dtu::Label, credits: u64) -> Result<(), Error> {
    log!(
        SYSC,
        "syscalls::create_sgate(dst={}, rgate={}, lbl={:#x}, credits={})",
        dst, rgate, label, credits
    );

    let req = syscalls::CreateSGate {
        opcode: syscalls::Operation::CREATE_SGATE.val,
        dst_sel: dst as u64,
        rgate_sel: rgate as u64,
        label: label,
        credits: credits,
    };
    send_receive_result(&req)
}

pub fn create_mgate(dst: Selector, addr: usize, size: usize, perms: Perm) -> Result<(), Error> {
    log!(
        SYSC,
        "syscalls::create_mgate(dst={}, addr={:#x}, size={:#x}, perms={:?})",
        dst, addr, size, perms
    );

    let req = syscalls::CreateMGate {
        opcode: syscalls::Operation::CREATE_MGATE.val,
        dst_sel: dst as u64,
        addr: addr as u64,
        size: size as u64,
        perms: perms.bits() as u64,
    };
    send_receive_result(&req)
}

pub fn create_rgate(dst: Selector, order: i32, msgorder: i32) -> Result<(), Error> {
    log!(
        SYSC,
        "syscalls::create_rgate(dst={}, order={}, msgorder={})",
        dst, order, msgorder
    );

    let req = syscalls::CreateRGate {
        opcode: syscalls::Operation::CREATE_RGATE.val,
        dst_sel: dst as u64,
        order: order as u64,
        msgorder: msgorder as u64,
    };
    send_receive_result(&req)
}

pub fn create_map(dst: Selector, vpe: Selector, mgate: Selector, first: Selector,
                  pages: u32, perms: Perm) -> Result<(), Error> {
    log!(
        SYSC,
        "syscalls::create_map(dst={}, vpe={}, mgate={}, first={}, pages={}, perms={:?})",
        dst, vpe, mgate, first, pages, perms
    );

    let req = syscalls::CreateMap {
        opcode: syscalls::Operation::CREATE_MAP.val,
        dst_sel: dst as u64,
        vpe_sel: vpe as u64,
        mgate_sel: mgate as u64,
        first: first as u64,
        pages: pages as u64,
        perms: perms.bits() as u64,
    };
    send_receive_result(&req)
}

pub fn create_vpe(dst: CapRngDesc, sgate: Selector, name: &str, pe: PEDesc,
                  sep: dtu::EpId, rep: dtu::EpId, tmuxable: bool) -> Result<PEDesc, Error> {
    log!(
        SYSC,
        "syscalls::create_vpe(dst={}, sgate={}, name={}, pe={:?}, sep={}, rep={}, tmuxable={})",
        dst, sgate, name, pe, sep, rep, tmuxable
    );

    let mut req = syscalls::CreateVPE {
        opcode: syscalls::Operation::CREATE_VPE.val,
        dst_crd: dst.value() as u64,
        sgate_sel: sgate as u64,
        pe: pe.value() as u64,
        sep: sep as u64,
        rep: rep as u64,
        muxable: tmuxable as u64,
        namelen: name.len() as u64,
        name: unsafe { intrinsics::uninit() },
    };

    // copy name
    for (a, c) in req.name.iter_mut().zip(name.bytes()) {
        *a = c as u8;
    }

    let reply: Reply<syscalls::CreateVPEReply> = send_receive(&req)?;
    match reply.data.error {
        0 => Ok(PEDesc::new_from(reply.data.pe as u32)),
        e => Err(Error::from(e as u32))
    }
}

pub fn derive_mem(dst: Selector, src: Selector, offset: usize, size: usize, perms: Perm) -> Result<(), Error> {
    log!(
        SYSC,
        "syscalls::derive_mem(dst={}, src={}, off={:#x}, size={:#x}, perms={:?})",
        dst, src, offset, size, perms
    );

    let req = syscalls::DeriveMem {
        opcode: syscalls::Operation::DERIVE_MEM.val,
        dst_sel: dst as u64,
        src_sel: src as u64,
        offset: offset as u64,
        size: size as u64,
        perms: perms.bits() as u64,
    };
    send_receive_result(&req)
}

pub fn vpe_ctrl(vpe: Selector, op: syscalls::VPEOp, arg: u64) -> Result<(), Error> {
    log!(
        SYSC,
        "syscalls::vpe_ctrl(vpe={}, op={:?}, arg={})",
        vpe, op, arg
    );

    let req = syscalls::VPECtrl {
        opcode: syscalls::Operation::VPE_CTRL.val,
        vpe_sel: vpe as u64,
        op: op.val,
        arg: arg as u64,
    };
    send_receive_result(&req)
}

pub fn vpe_wait(vpes: &[Selector]) -> Result<(Selector, i32), Error> {
    log!(
        SYSC,
        "syscalls::vpe_wait(vpes={})",
        vpes.len()
    );

    let mut req = syscalls::VPEWait {
        opcode: syscalls::Operation::VPE_WAIT.val,
        vpe_count: vpes.len() as u64,
        sels: unsafe { intrinsics::uninit() },
    };
    for i in 0..vpes.len() {
        req.sels[i] = vpes[i] as u64;
    }

    let reply: Reply<syscalls::VPEWaitReply> = send_receive(&req)?;
    match reply.data.error {
        0 => Ok((reply.data.vpe_sel as Selector, reply.data.exitcode as i32)),
        e => Err(Error::from(e as u32))
    }
}

pub fn exchange(vpe: Selector, own: CapRngDesc, other: Selector, obtain: bool) -> Result<(), Error> {
    log!(
        SYSC,
        "syscalls::exchange(vpe={}, own={}, other={}, obtain={})",
        vpe, own, other, obtain
    );

    let req = syscalls::Exchange {
        opcode: syscalls::Operation::EXCHANGE.val,
        vpe_sel: vpe as u64,
        own_crd: own.value(),
        other_sel: other as u64,
        obtain: obtain as u64,
    };
    send_receive_result(&req)
}

pub fn delegate(vpe: Selector, sess: Selector, crd: CapRngDesc,
                args: &mut syscalls::ExchangeArgs) -> Result<(), Error> {
    log!(SYSC, "syscalls::delegate(vpe={}, sess={}, crd={})", vpe, sess, crd);

    exchange_sess(vpe, syscalls::Operation::DELEGATE, sess, crd, args)
}

pub fn obtain(vpe: Selector, sess: Selector, crd: CapRngDesc,
              args: &mut syscalls::ExchangeArgs) -> Result<(), Error> {
    log!(SYSC, "syscalls::obtain(vpe={}, sess={}, crd={})", vpe, sess, crd);

    exchange_sess(vpe, syscalls::Operation::OBTAIN, sess, crd, args)
}

fn exchange_sess(vpe: Selector, op: syscalls::Operation, sess: Selector, crd: CapRngDesc,
                 args: &mut syscalls::ExchangeArgs) -> Result<(), Error> {
    let req = syscalls::ExchangeSess {
        opcode: op.val,
        vpe_sel: vpe as u64,
        sess_sel: sess as u64,
        crd: crd.value(),
        args: args.clone(),
    };

    let reply: Reply<syscalls::ExchangeSessReply> = send_receive(&req)?;
    if reply.data.error == 0 {
        *args = reply.data.args;
    }

    match reply.data.error {
        0 => Ok(()),
        e => Err(Error::from(e as u32))
    }
}

pub fn revoke(vpe: Selector, crd: CapRngDesc, own: bool) -> Result<(), Error> {
    log!(
        SYSC,
        "syscalls::revoke(vpe={}, crd={}, own={})",
        vpe, crd, own
    );

    let req = syscalls::Revoke {
        opcode: syscalls::Operation::REVOKE.val,
        vpe_sel: vpe as u64,
        crd: crd.value(),
        own: own as u64,
    };
    send_receive_result(&req)
}

pub fn forward_write(mgate: Selector, data: &[u8], off: usize,
                     flags: syscalls::ForwardMemFlags, event: u64) -> Result<(), Error> {
    log!(
        SYSC,
        "syscalls::forward_write(mgate={}, count={}, off={}, flags={}, event={})",
        mgate, data.len(), off, flags.bits(), event
    );

    let mut req = syscalls::ForwardMem {
        opcode: syscalls::Operation::FORWARD_MEM.val,
        mgate_sel: mgate as u64,
        offset: off as u64,
        flags: (flags | syscalls::ForwardMemFlags::WRITE).bits() as u64,
        event: event as u64,
        len: data.len() as u64,
        data: unsafe { intrinsics::uninit() },
    };
    req.data[0..data.len()].copy_from_slice(data);

    send_receive_result(&req)
}

pub fn forward_read(mgate: Selector, data: &mut [u8], off: usize,
                    flags: syscalls::ForwardMemFlags, event: u64) -> Result<(), Error> {
    log!(
        SYSC,
        "syscalls::forward_read(mgate={}, count={}, off={}, flags={}, event={})",
        mgate, data.len(), off, flags.bits(), event
    );

    let req = syscalls::ForwardMem {
        opcode: syscalls::Operation::FORWARD_MEM.val,
        mgate_sel: mgate as u64,
        offset: off as u64,
        flags: flags.bits() as u64,
        event: event as u64,
        len: data.len() as u64,
        data: unsafe { intrinsics::uninit() },
    };

    let reply: Reply<syscalls::ForwardMemReply> = send_receive(&req)?;
    if reply.data.error == 0 {
        let len = data.len();
        data.copy_from_slice(&reply.data.data[0..len]);
    }

    match reply.data.error {
        0 => Ok(()),
        e => Err(Error::from(e as u32))
    }
}

pub fn noop() -> Result<(), Error> {
    let req = syscalls::Noop {
        opcode: syscalls::Operation::NOOP.val,
    };
    send_receive_result(&req)
}

pub fn exit(code: i32) {
    log!(
        SYSC,
        "syscalls::exit(code={})",
        code
    );

    let req = syscalls::VPECtrl {
        opcode: syscalls::Operation::VPE_CTRL.val,
        vpe_sel: 0,
        op: syscalls::VPEOp::STOP.val,
        arg: code as u64,
    };
    send(&req).unwrap();
}
